### PR TITLE
check if libyaml is already present before building it

### PR DESCRIPTION
**Pull request by alsora**
This PR addresses https://github.com/ros2/libyaml_vendor/issues/40

Signed-off-by: Alberto Soragna <alberto.soragna@gmail.com>

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -14,8 +14,16 @@ if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
   add_compile_options(-Wall -Wextra -Wconversion -Wno-sign-conversion -Wpedantic -Wnon-virtual-dtor -Woverloaded-virtual)
 endif()
 
+option(FORCE_BUILD_VENDOR_PKG
```
Comment (hidmic) [inline CMakeLists.txt:17]: @alsora nit: 
```suggestion
option(FORCE_BUILD_VENDORED_LIBYAML
```

or anything along those lines. `FORCE_BUILD_VENDOR_PKG` is a bit generic and could span several packages.

File: CMakeLists.txt [hunk 2]
```diff
@@ -84,7 +92,18 @@ macro(build_libyaml)
   set(yaml_LIBRARIES yaml)
 endmacro()
 
-build_libyaml()
+# Skip building yaml if the expected version is already present in the system
+if (yaml_FOUND)
+  if("${yaml_VERSION}" VERSION_EQUAL 0.2.5)
+    set(_SKIP_YAML_BUILD 1)
+  else()
+    message(WARNING "A wrong version of libyaml is already present in the system: ${yaml_VERSION}")
```
Comment (hidmic) [inline CMakeLists.txt:100]: @alsora shall we add that it will be ignored?

**Review by hidmic — COMMENTED** [review]
Makes sense to me.

File: CMakeLists.txt (hunk 1 above)
Comment (alsora) [reply CMakeLists.txt:17]: Having a generic name was done on purpose.
This is the same name as used in other vendor packages
https://github.com/ros2/spdlog_vendor/blob/master/CMakeLists.txt#L6
https://github.com/ros2/yaml_cpp_vendor/blob/master/CMakeLists.txt#L75
https://github.com/ros2/console_bridge_vendor/blob/master/CMakeLists.txt#L16

File: CMakeLists.txt (hunk 2 above)
Comment (alsora) [reply CMakeLists.txt]: Done

File: CMakeLists.txt (hunk 1 above)
Comment (hidmic) [reply CMakeLists.txt:17]: Good point. Nevermind.

File: CMakeLists.txt [hunk 3]
```diff
@@ -93,19 +93,21 @@ macro(build_libyaml)
...
 # Skip building yaml if the expected version is already present in the system
-if (yaml_FOUND)
+if(yaml_FOUND)
   if("${yaml_VERSION}" VERSION_EQUAL 0.2.5)
     set(_SKIP_YAML_BUILD 1)
   else()
-    message(WARNING "A wrong version of libyaml is already present in the system: ${yaml_VERSION}")
+    message(WARNING
+      "A wrong version of libyaml is already present in the system: ${yaml_VERSION}."
+      "It will be ignored and the expected version will be built.")
```
Comment (hidmic) [inline CMakeLists.txt:102]: @alsora nit: let's mention what the expected version is.

**Review by hidmic — APPROVED** [review]
LGTM pending green CI

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15228/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9948/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12908/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15411/)

**Comment by alsora** [comment]
Sorry for the force push.. used wrong account without sign off...

File: CMakeLists.txt (hunk 3 above)
Comment (alsora) [reply CMakeLists.txt]: I hard-coded a 0.2.5 in the message.
The number is used all over the file, but I think that using a variable would make the CMakelists much less readable.

**Review by hidmic — APPROVED** [review]
Code is otherwise the same. @nuclearsandwich @cottsay do you foresee any fallout after this in other platforms like Debian or RHEL?

File: CMakeLists.txt [hunk 4]
```diff
@@ -84,8 +92,21 @@ macro(build_libyaml)
...
 endmacro()
 
-build_libyaml()
-set(extra_test_dependencies libyaml-0.2.5)
+# Skip building yaml if the expected version is already present in the system
+if(yaml_FOUND)
+  if("${yaml_VERSION}" VERSION_EQUAL 0.2.5)
+    set(_SKIP_YAML_BUILD 1)
+  else()
+    message(WARNING
```
Comment (cottsay) [inline CMakeLists.txt:100]: I'm not sure we want this to be a warning. An info-level message is probably sufficient.

File: CMakeLists.txt [hunk 5]
```diff
@@ -84,8 +92,21 @@ macro(build_libyaml)
   set(yaml_LIBRARIES yaml)
 endmacro()
 
-build_libyaml()
-set(extra_test_dependencies libyaml-0.2.5)
+# Skip building yaml if the expected version is already present in the system
+if(yaml_FOUND)
+  if("${yaml_VERSION}" VERSION_EQUAL 0.2.5)
```
Comment (cottsay) [inline CMakeLists.txt:97]: Maybe `VERSION_GREATER_EQUAL` would be better - that's what we're doing in other vendor packages, I think.

**Review by cottsay — CHANGES_REQUESTED** [review]
If we're going to do this (and I see no reason not to), we should make it behave consistently by taking a dependency on the libyaml-dev rosdep key in the package manifest. Opportunistically using the system package might make debugging harder.

File: CMakeLists.txt (hunk 4 above)
Comment (alsora) [reply CMakeLists.txt:100]: My first idea was for this to be FATAL... the reason being that if you have libyaml in your system it means that you are already using it and having two versions in your system is the recipe for a disaster.

This obviously is geared towards a real robotics application use-case.
If you run ROS 2 on your laptop then you will definitely have packages installed that you don't use.

File: CMakeLists.txt (hunk 5 above)
Comment (alsora) [reply CMakeLists.txt:97]: Yes other vendor packages do that.
However it was pointed out that here we need exact version.

Comment (hidmic) [reply CMakeLists.txt:97]: IIUC version pegging is what enables us to claim this is a QL1 package. API/ABI compatibility is not guaranteed by `libyaml`.

**Comment by Tobias-Fischer** [comment]
Hi all,

What is the status of this PR? It would be great to see it merged in.

File: CMakeLists.txt [hunk 6]
```diff
@@ -14,8 +14,16 @@ if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
...
 endif()
 
+option(FORCE_BUILD_VENDOR_PKG
+  "Build libyaml from source, even if system-installed package is available"
+  OFF)
+
 find_package(ament_cmake REQUIRED)
 
+if(NOT FORCE_BUILD_VENDOR_PKG)
+  find_package(yaml QUIET)
```
[the remark this answers is not in the document]
Comment (traversaro) [reply CMakeLists.txt:24]: I modified this PR to correctly handle the case in which yaml is installed in the system via pkg-config, se https://github.com/ros2/libyaml_vendor/pull/45 .

**Comment by hidmic** [comment]
I'll close this PR as it's been augmented by #45. @alsora feel free to re-open as needed.

**Pull request closed without merging**